sim/src/main/cc/bridges/UARTModule.cc

```diff
@@ -0,0 +1,75 @@
+// See LICENSE for license details.
+
+#ifndef RTLSIM
+#include "simif_f1.h"
+#define SIMIF simif_f1_t
+#else
+#include "simif_emul.h"
+#define SIMIF simif_emul_t
+#endif
+
+#include "bridges/uart.h"
+#include "simif_peek_poke.h"
+
+class UARTModuleTest final : public simif_peek_poke_t {
+public:
+  class Handler final : public uart_handler {
+  public:
+    ~Handler() {
+      // Check that the input and output buffers are equal.
+      if (in_buffer != out_buffer) {
+        fprintf(stderr,
+                "Buffer mismatch:\n  %s\n  %s\n",
+                in_buffer.c_str(),
+                out_buffer.c_str());
+        abort();
+      }
+    }
+
+    std::optional<char> get() override {
+      if (in_index >= in_buffer.size())
+        return std::nullopt;
+      return in_buffer[in_index++];
+    }
+
+    void put(char data) override { out_buffer += data; }
+
+  private:
+    const std::string in_buffer = "We are testing the UART bridge";
+    std::string out_buffer;
+    size_t in_index = 0;
+  };
+
+  UARTModuleTest(simif_t &simif)
+      : simif_peek_poke_t(&simif, PEEKPOKEBRIDGEMODULE_0_substruct_create),
+        uart(std::make_unique<uart_t>(&simif,
+                                      UARTBRIDGEMODULE_0_substruct_create,
+                                      std::make_unique<Handler>())) {}
+
+  void run() {
+    // Reset the device.
+    poke(reset, 1);
+    step(1);
+    poke(reset, 0);
+    step(1);
+
+    // Tick for a while.
+    for (unsigned i = 0; i < 30; ++i) {
```

Comment:
How long does this test take to run? For the string you're writing you could make this test run much more quickly by calling tick more frequently relative to step. For testing bridges like this, you can favor the non-blocking variant of "step" and instead write:

```
step(<some_reasonable_upper_bound>, false)
while(!done() <(&& possibly other conditions)>) {
  uart->tick();
}
```

Reply:
Basically the UART is sitting idle for the bulk of each of the 10K cycle steps.

Reply:
Thanks for clarifying this!